### ISSUE TITLE
`confirm_quit` has no effect

### ISSUE THREAD

**Issue by yuri1969** [zed-industries/zed#14439]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

1. Set `"confirm_quit": true` in `settings.json`.
1. Close Zed
1. Zed closes with no prompt

Using **i3wm**, but recent Ubuntu behaves the same.

### Environment

Zed: v0.143.7 (Zed)
OS: Linux X11 gentoo 2.15
Memory: 31.2 GiB
Architecture: x86_64

### If applicable, attach your Zed.log file to this issue.

<details><summary>Zed.log</summary><pre>
2024-07-12T21:56:52+02:00 [ERROR] oneshot canceled
2024-07-12T21:56:52+02:00 [ERROR] oneshot canceled
2024-07-12T21:56:52+02:00 [ERROR] window not found
2024-07-12T21:56:52+02:00 [ERROR] window not found

**Comment by yuri1969** [comment]
Still applies to Zed 0.178.5 7772fae131a6ea71b416373eb744825f4cbe83ff.

**Comment by TroySigX** [comment]
It works when we type `ctrl+q`, but doesn't work we click the close button or `alt-f4`. Might be helpful to the devs.

**Comment by mgsloan** [comment]
I think this is related to Zed's window close behavior on all platforms, so not Linux specific.

The current behavior of `confirm_quit: true` is to confirm when the quit action or reload workspace action is invoked.  It's not clear to me whether external window close requests should also involve this confirmation when this is set.  I'm guessing the motivation for the setting is because those actions may be bound to keyboard shortcuts which may be accidentally typed.

Should confirmation be asked for every window close even if there are no unsaved files?  If so that's more about confirming window closing than confirming application quit.

Such a change would go in `Workspace::prepare_to_close` or related functions.  It would need to track whether some other form of confirmation dialog has been displayed, and only display this one if no other confirmation was asked.

**Comment by TroySigX** [comment]
Thanks for the explanation!

So the current behavior is: 
- The `:quit` command that closes a window will show a confirmation popup when there's an unsaved file.
- The `zed: quit` command that closes the whole zed application does not, whether there's still unsaved files or not

For me, it's okay to not show the confirmation popup when there's no unsaved files, but it should be shown when we close any unsaved file or when we close the whole app while there's still unsaved files. The whole point of the confirmation is to make sure that we want those changes to be saved on disk. With the current behavior, closing the zed app when there's unsaved files will automatically save those changes on disk without anyway to recover to the previous version, which is very dangerous in cases where we may change the code to debug and close the app without remembering all those changes or something like that.

**Comment by mgsloan** [comment]
@TroySigX Ah, it would indeed be a bug if you can close a window with unsaved files and there is no confirmation, and that happens regardless of the setting for `confirm_quit`.  For me on debian + x11 + xmonad it does ask for confirmation when I close windows with unsaved files using xmonad, because it's sending `WM_DELETE_WINDOW`.

Which desktop environment / window manager are you using?  Do you have a setting for `autosave`?

**Comment by TroySigX** [comment]
I don't have `autosave` enabled. Here's my system:
```
Zed: v0.192.8 (Zed) 
OS: Linux Wayland endeavouros unknown
Memory: 7.6 GiB
Architecture: x86_64
GPU: NVIDIA GeForce GTX 1050 || NVIDIA || 575.64
```

If it helps, I'm using Arch Linux (Endeavour OS) with KDE.

**Comment by mgsloan** [comment]
Thanks for the info!  I think I may have tracked it down.  The current behavior is to not confirm closing if workspace persistence succeeds in persisting the unsaved file states, which does seem a bit surprising to me.  Makes it more likely that frictions like those described in #21528 occur.

If you create a new file that has no path (untitled), and edit it, that creates a state where confirmation will happen on close since files like that don't currently support being serialized as part of workspace persistence.  Does confirmation occur in that case?

**Comment by TroySigX** [comment]
Hmm, I'm not sure if I can edit a new file without naming it first.

![Image](https://github.com/user-attachments/assets/c11dc247-2b99-4eab-92c2-5d67e974b52e)

After I name it and edit the file, closing zed doesn't show any confirmation popup.

**Comment by TroySigX** [comment]
@mgsloan Okay I managed to do what you asked. Doing the command `zed: quit` (as well as typing `alt-f4`) still doesn't show any confirmation popup. Though when reopening zed, the unsaved panel persists without saving anything on disk

https://github.com/user-attachments/assets/c8f9a1e3-e275-4d8b-908b-bd53f2227012

**Comment by m-ar-c** [comment]
`zed: quit` does ask for a confirmation, but system close window function (`Alt+F4`) does not. (Meaning I can accidentally permanently lose external acp agent ongoing conversation (#37074) for example).

Debian GNU/Linux forky/sid
Wayland Gnome 49
Zed 0.210.4

**Comment by aizigao** [comment]
same issue in 
Zed Preview 1.2.2 
Commit: 1666c0bd811f52ece59300c811afac6b7eae76d0
Version: 1.2.2+preview.267.1666c0bd811f52ece59300c811afac6b7eae76d0

**Comment by andrevmatos** [comment]
yes, this still happening also on MacOS; even with `"confirm_quit": true`, a cmd+shift+W closes without confirmation, losing acp and terminal state. Can we please reopen this issue?
